ocsp/responder/multi_source.go

```diff
@@ -54,50 +66,90 @@ func (src *multiSource) Response(ctx context.Context, req *ocsp.Request) (*Respo
 	// passes these checks, return its response instead.
 	select {
 	case <-ctx.Done():
-		src.counter.WithLabelValues("timed_out").Inc()
-		return nil, fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+		src.counter.WithLabelValues("primary_timed_out").Inc()
+		return nil, ctx.Err()
+
+	case r := <-primaryChan:
+		// If there was an error requesting from the primary, don't bother
+		// waiting for the secondary, because we wouldn't be able to
+		// check the secondary's status against the (more reliable) primary's
+		// status.
+		if r.err != nil {
+			src.counter.WithLabelValues("primary_error").Inc()
+			return nil, r.err
+		}
+		primaryResponse = r.resp
+	}
 
-	case primaryResult := <-primaryChan:
+	// The primary response was fresh enough to serve, go ahead and serve it.
+	if time.Since(primaryResponse.ThisUpdate) < src.expectedFreshness {
+		src.checkSecondary(primaryResponse, secondaryChan)
 		src.counter.WithLabelValues("primary_result").Inc()
-		return primaryResult.resp, primaryResult.err
+		return primaryResponse, nil
+	}
+
+	// The primary response was too stale to (ideally) serve. This will be
+	// a common path once we stop ocsp-updater from writing updated blobs
+	// to MariaDB. Try to serve from the secondary.
+	var secondaryResponse *Response
+	select {
+	case <-ctx.Done():
+		src.counter.WithLabelValues("timed_out_awaiting_secondary").Inc()
+		// Best-effort: return the primary response even though it's stale.
+		return primaryResponse, nil
 
 	case secondaryResult := <-secondaryChan:
-		// If secondary returns first, wait for primary to return for
-		// comparison.
-		var primaryResult responseResult
+		if secondaryResult.err != nil {
+			if errors.Is(secondaryResult.err, rocsp.ErrRedisNotFound) {
+				// This case will happen for several hours after first issuance.
```

Comment:
This case _may_ happen approximately one ocsp-updater cycle after fresh issuance, if ExpectedFreshness is too low, right? But if ExpectedFreshness is set to MinTimeToExpiry+Window as suggested, then a new response should be generated before we start considering the original response stale, and we _should_ never hit this case. I think?

Reply:
Remember that on new issuance we write to MariaDB but not to Redis. So based on the current ocsp-updater algorithm, we would actually expect this to happen for 60h (MinTimeToExpiry) after first issuance, since that's how long ocsp-updater would take to come along and decide to update (and that update would include a write to Redis).

Now, we're going to have to change the ocsp-updater algorithm once we're not writing to MariaDB. Part of that will probably include a scan for recently-issued certificates with no entry in Redis, which would make this case happen much less frequently.

Reply:
Ah right!